### PR TITLE
Message forwarding

### PR DESCRIPTION
**Pull request by aforge**
This is relatively raw and needs more work.
Sending out early for feedback.

### CODE REVIEW

File: src/views/messages-view.jsx [hunk 1]
```diff
@@ -665,14 +676,20 @@ class MessagesView extends React.Component {
   // sendMessage sends the message with an optional subscription to topic first.
   sendMessage(msg, uploadCompletionPromise, uploader) {
     let head;
-    if (this.state.reply && this.state.reply.content) {
+    if (this.props.forwardedMessage) {
+      // We are forwarding a message.
+      msg = this.props.forwardedMessage.msg;
+      head = this.props.forwardedMessage.head;
```
Comment (or-else) [inline src/views/messages-view.jsx:682]: Please add a header per spec: https://github.com/tinode/chat/blob/master/docs/API.md#pub

File: src/views/tinode-web.jsx [hunk 2]
```diff
@@ -1428,6 +1445,32 @@ class TinodeWeb extends React.Component {
     });
   }
 
+  handleShowForwardMenu(params) {
+    if (this.state.sidePanelSelected == 'newtpk') {
+      // If the Find panel is active, close it.
+      this.handleSidepanelCancel();
+    }
+    const header = "Forwarded from " + params.userName;
```
Comment (or-else) [inline src/views/tinode-web.jsx:1453]: Internationalize?

Comment (or-else) [reply src/views/tinode-web.jsx:1453]: Strip previous Forwarded for header?
Screenshot 2021-09-21 at 11 16 51

**Review by or-else — APPROVED** [review]
Pretty good. I'll tweak the visuals. Also this:
Screenshot 2021-09-21 at 11 22 29

**Review by or-else — APPROVED** [review]

File: src/views/messages-view.jsx (hunk 1 above)
Comment (aforge) [reply src/views/messages-view.jsx]: Done.

File: src/lib/formatters.js [hunk 3]
```diff
@@ -393,3 +393,22 @@ export function replyFormatter(style, data, values, key) {
   values = [React.createElement(LazyImage, attr, null), ' ', attr.alt];
   return React.createElement(React.Fragment, {key: key}, values);
 }
+
+// Creates a "forward" icon fragment.
+export function forwardedDecorator(context) {
+  return [<i key="fwd" className="material-icons">forward</i>];
```
[the remark this answers is not in the document]
Comment (aforge) [reply src/lib/formatters.js]: Thanks. Changed.

**Comment by aforge** [comment]
Stripping the "Forwarded from" mention from originally forwarded messages took an API change: https://github.com/tinode/tinode-js/pull/49

**Comment by or-else** [comment]
The Forward To menu should not list topics without the `W` access (or list them as disabled). Otherwise LGTM. 
I would tweak some visuals though.

**Comment by or-else** [comment]
Screenshot 2021-09-26 at 10 00 34

Maybe Remove `Alice Johnson` here leaving just ➦. Or simply remove the whole MN if removing just the name is hard.

**Comment by aforge** [comment]
> Screenshot 2021-09-26 at 10 00 34
> 
> Maybe Remove `Alice Johnson` here leaving just ➦. Or simply remove the whole MN if removing just the name is hard.

Done.

**Comment by aforge** [comment]
> The Forward To menu should not list topics without the `W` access (or list them as disabled). Otherwise LGTM. I would tweak some visuals though.

Done.

**Pull request merged**